psi4/driver/p4util/solvers.py

```diff
@@ -310,3 +310,159 @@ def extrapolate(self, out=None):
             out.axpy(c, self.state[num])
 
         return out
+
+
+# Generalized Davidson solver
+
+def davidson_solver(ax_function, preconditioner, guess, e_conv=1.0E-8,
+        r_conv=None, no_eigs=1, max_vecs_per_root=20, maxiter=100):
+    """
+    Solves for the lowest few eigenvalues and eigenvectors of the given real symmetric matrix
+
+    Parameters
+    -----------
+    ax_function : function
```

Comment:
What's the advantage of having the user pass in a function to perform matrix multiplication, as opposed to having the user pass in the matrix and letting the program do the multiplication?

Reply:
A CI matrix can be a billion x billion. You don't want to form that. Instead, you contract directly.

Reply:
Same thing for SCF stability analysis and TD-DFT. Say you have 100 occupied orbitals, and 1000 virtual ones. Then your matrix dimension will be 100*1000 = 100 000 x 100 000. This will take about 80 GB of memory, whereas even 20 trial vectors only take 16 MB.